### PR TITLE
expression: create a new aggregate distinct if not.

### LINKED ISSUE #324: Test bug seems in groupby.
```
DROP TABLE IF EXISTS tab1;
CREATE TABLE tab1(col0 INTEGER, col1 INTEGER, col2 INTEGER);
INSERT INTO tab1 VALUES(22,6,8);
INSERT INTO tab1 VALUES(28,57,45);
INSERT INTO tab1 VALUES(82,44,71);

TiDB got,
SELECT - col0 + - - 20 * - + col0 * 25 * + col2 - 72 FROM tab1 GROUP BY col2, col0 HAVING - AVG ( DISTINCT 59 ) NOT BETWEEN 36 AND - col0 + - 57 * col0;
+-----------------------------------------+
| -col0 + --20 * -+col0 * 25 * +col2 - 72 |
+-----------------------------------------+
| -88094                                  |
+-----------------------------------------+

But MySQL got,
+-----------------------------------------------+
| - col0 + - - 20 * - + col0 * 25 * + col2 - 72 |
+-----------------------------------------------+
|                                        -88094 |
|                                       -630100 |
|                                      -2911154 |
+-----------------------------------------------+
```

### PR DESCRIPTION
**Pull request by siddontang**
Fix #324

### CODE REVIEW

File: expression/call.go [hunk 1]
```diff
@@ -126,15 +126,17 @@ func (c *Call) Eval(ctx context.Context, args map[interface{}]interface{}) (v in
...
-		c.d = builtin.CreateAggregateDistinct(c.F, c.Distinct)
-	}
-
 	if args != nil {
 		args[builtin.ExprEvalFn] = c
 		args[builtin.ExprEvalArgCtx] = ctx
-		args[builtin.ExprAggDistinct] = c.d
+		aggDistinct, ok := args[c.distinctKey]
+		if !ok {
+			// create a aggregate distinct if not.
```
Comment (qiuyesuifeng) [inline expression/call.go]: create an aggregate

**Comment by qiuyesuifeng** [comment]
LGTM

File: expression/call.go [hunk 2]
```diff
@@ -40,7 +40,7 @@ type Call struct {
 	// so we can ignore it in other functions
 	Distinct bool
 
-	d *builtin.AggregateDistinct
+	distinctKey *int
```
Comment (shenli) [inline expression/call.go]: I am confused about this. Why change to *int?

Comment (siddontang) [inline expression/call.go]: To generate an unique key using new(int).

Comment (shenli) [inline expression/call.go]: Why not use the address of Call as unique key?

Comment (siddontang) [inline expression/call.go]: Because inner already uses it to store aggregate result.

Comment (shenli) [inline expression/call.go]: OK. But please add some comments. We are all not clear about this.

**Comment by shenli** [comment]
LGTM

**Pull request merged**